### DISCUSSION TITLE
Assistance with solution for ERR_UNKNOWN_URL_SCHEME ?

### DISCUSSION THREAD

**Discussion by bliqqo** [Voltstro-Studios/UnityWebBrowser#142] [Q&A]
Hello!

I've encountered one issue regarding the use of Zoom's website (Video Conferencing) on this package. Upon create or joining meetings (zoom us -> login -> main page -> Join / Host ), Zoom redirects leads to an ERR_UNKNOWN_URL_SCHEME, not something I've seen when using Zoom before. 

![image](https://user-images.githubusercontent.com/62570856/215208986-e1554022-2b04-4d13-b31d-edd91182e737.png)

I've observed successful webcam support through this package as well as successful meetings on Zoom by altering the URL- to direct the browser specifically to the web meetings. ( Referenced here: https://magpcss.org/ceforum/viewtopic.php?f=6&t=19216 )

I would like to know if there is any direction someone might have to prevent this error. Might it be resolved with a newer version update to CEF? CEFsharp? CEFGlue?

**Comment by Voltstro** [comment]
It's because of the custom URL scheme (`zoommtg://`). Apps can register custom URL schemes with the OS to tell browsers then when a URL with that scheme is accessed, to open the app itself. By default, CEF blocks all of these calls (for security reasons).

You can tell CEF to allow an unknown URL scheme to attempt execution via the registered OS protocol handler. This is done via the [`CefResourceRequestHandler.OnProtocolExecution`](https://gitlab.com/Voltstro-Studios/WebBrowser/cefglue/-/blob/master/CefGlue/Classes.Handlers/CefResourceRequestHandler.cs#L265), so you will need to override the handler in UWB's CEF engine (as it's not currently used), and check for allowed URL schemes, then set `allowOSExecution` to `true`.

**Reply by bliqqo** [reply]
Thank you for the explanation.

I have followed your advice, but I'm struggling to understand how to reintegrate these changes back into this project. Relating to CEF and in the context of UWB's CEF file structure specifically. 

From what I understand I need to recompile CEF, and replace the package files from the repo with the custom binaries and files.
<img src="https://user-images.githubusercontent.com/62570856/215906240-f71a779e-a04e-42de-a641-38271ae6d202.png" data-canonical-src="https://user-images.githubusercontent.com/62570856/215905559-3fad02e8-5562-4011-97cc-7f92f2c26a48.png" width="250" height="200" />

It's very unclear to me how to actually accomplish this, I've attempted it with 'CMaker' and 'automate-git.py'. I've not been able to find a straigtforward path in either direction, but being as there has been similar discussion: https://github.com/Voltstro-Studios/UnityWebBrowser/discussions/111

Would the proposed changes be reintegrated the same way?

**Reply by Voltstro** [reply]
You don't need to recompile CEF for this, it just creating a custom handler class implementing CefResourceRequestHandler. Create a new class in the UnityWebBrowser.Engine.Cef project, implement CefResourceRequestHandler, override OnProtocolExecution and make [UwbCefClient](https://github.com/Voltstro-Studios/UnityWebBrowser/blob/master/src/UnityWebBrowser.Engine.Cef/Browser/UwbCefClient.cs) aware of the custom handler.

Have a look at the other handlers in [UnityWebBrowser.Engine.Cef](https://github.com/Voltstro-Studios/UnityWebBrowser/tree/master/src/UnityWebBrowser.Engine.Cef/Browser) for an example.

**Reply by bliqqo** [reply]
In reference to [UwbCefClient](https://github.com/Voltstro-Studios/UnityWebBrowser/blob/6ced6c2b798feee5ca3578eb67f052441b8a2320/src/UnityWebBrowser.Engine.Cef/Browser/UwbCefClient.cs), There isn't an equivalent to replace a CEFResorceRequestHandler; only saving the new handler as a local variable I assume wouldn't be sufficient, am I missing something?

Also, All of these files from the /src/ directory, aren't part of the unity runtime scripts; I haven't confirmed how else these files are recompiled / repacked. CEF recompilation isn't necessary (_thankfully_), but what steps are needed to package files such as to UwbCefClient & any new handlers into a unity project?

powershell scripts? such as [publish-cef-engine-windows.ps1](https://github.com/Voltstro-Studios/UnityWebBrowser/blob/6ced6c2b798feee5ca3578eb67f052441b8a2320/src/DevScripts/publish-cef-engine-windows.ps1)?
My test using this hasn't yet been fruitful.

Thanks again for all the assistance & direction, has helped tremendously

**Reply by Voltstro** [reply]
Ok I just had a look, the CefResourceRequestHandler is provided by the `CefRequestHandler`. UWB's Cef engine [already has one](https://github.com/Voltstro-Studios/UnityWebBrowser/blob/6ced6c2b798feee5ca3578eb67f052441b8a2320/src/UnityWebBrowser.Engine.Cef/Browser/UwbCefRequestHandler.cs), so you will just need to provide the class implementing CefResourceRequestHandler in the `GetResourceRequestHandler` method. Also have a good read of the method's XML doc comment.

---

The `src` folder holds all of UWB's source code and devtools. There are engine and shared projects which are external to Unity (Regular .NET projects). And there is the Unity Packages.

The Unity packages them selves are located in [`src/Packages`](https://github.com/Voltstro-Studios/UnityWebBrowser/tree/master/src/Packages). The DevScripts are used to build the external projects and copy the externally built blobs to their packages. So yes, if you want to build the external CEF engine project, run [publish-cef-engine-windows.ps1](https://github.com/Voltstro-Studios/UnityWebBrowser/blob/6ced6c2b798feee5ca3578eb67f052441b8a2320/src/DevScripts/publish-cef-engine-windows.ps1).

**Reply by bliqqo** [reply]
>  UWB's Cef engine [already has one](https://github.com/Voltstro-Studios/UnityWebBrowser/blob/6ced6c2b798feee5ca3578eb67f052441b8a2320/src/UnityWebBrowser.Engine.Cef/Browser/UwbCefRequestHandler.cs), so you will just need to provide the class implementing CefResourceRequestHandler in the `GetResourceRequestHandler` method. Also have a good read of the method's XML doc comment.

After making the proposed changes:

1.    I used  [publish-cef-engine-windows.ps1](https://github.com/Voltstro-Studios/UnityWebBrowser/blob/6ced6c2b798feee5ca3578eb67f052441b8a2320/src/DevScripts/publish-cef-engine-windows.ps1) to  set up for a build for use in my Unity Project. (along with updating Visual Studio builds of the project)
2.    In unity editor, via package manager I added a new package from disk ([.../src/Packages/UnityWebBrowser.Engine.Cef.Win-x64/](https://github.com/Voltstro-Studios/UnityWebBrowser/tree/master/src/Packages/UnityWebBrowser.Engine.Cef.Win-x64))

After attempting to run it, Currently it fails to start with the following errors (in that order):

<details>
  <summary>
    <b>[Web Browser]: [Engine]: Uncaught exception occured in early init!</b>
  </summary><pre>
  [Web Browser]: [Engine]: Uncaught exception occured in early init!
System.DllNotFoundException: Dll was not found.
     at Xilium.CefGlue.Interop.libcef.api_hash(Int32 entry)
     at Xilium.CefGlue.CefRuntime.CheckVersionByApiHash()
     at Xilium.CefGlue.CefRuntime.CheckVersion()
     at Xilium.CefGlue.CefRuntime.Load(String path)
     at Xilium.CefGlue.CefRuntime.Load()
     at UnityWebBrowser.Engine.Cef.Core.CefEngineControlsManager..ctor()
     at UnityWebBrowser.Engine.Cef.Core.UwbCefEngineEntry.EarlyInit(LaunchArguments launchArguments, String[] args)  
     at VoltstroStudios.UnityWebBrowser.Engine.Shared.Core.EngineEntryPoint.<>c__DisplayClass15_0.b__19(LaunchArguments parsedArgs)
UnityEngine.Logger:LogError (string,object)
VoltstroStudios.UnityWebBrowser.Logging.DefaultUnityWebBrowserLogger:Error (object) (at Library/PackageCache/dev.voltstro.unitywebbrowser@2.0.1/Runtime/Logging/DefaultUnityWebBrowserLogger.cs:36)
VoltstroStudios.UnityWebBrowser.Logging.ProcessLogHandler:HandleOutputProcessLog (object,System.Diagnostics.DataReceivedEventArgs) (at Library/PackageCache/dev.voltstro.unitywebbrowser@2.0.1/Runtime/Logging/ProcessLogHandler.cs:54)
System.Threading._ThreadPoolWaitCallback:PerformWaitCallback ()
</pre>
</details>

<details>
  <summary>
    <b>[Web Browser]: The engine did not get ready within engine startup timeout! The engine process is not even running! Exit code: -1.</b>
  </summary><pre>
 [Web Browser]: The engine did not get ready within engine startup timeout! The engine process is not even running! Exit code: -1.
UnityEngine.Logger:LogError (string,object)
VoltstroStudios.UnityWebBrowser.Logging.DefaultUnityWebBrowserLogger:Error (object) (at Library/PackageCache/dev.voltstro.unitywebbrowser@2.0.1/Runtime/Logging/DefaultUnityWebBrowserLogger.cs:36)
VoltstroStudios.UnityWebBrowser.Core.WebBrowserClient/<WaitForEngineReadyTask>d__57:MoveNext () (at Library/PackageCache/dev.voltstro.unitywebbrowser@2.0.1/Runtime/Core/WebBrowserClient.cs:415)
Cysharp.Threading.Tasks.CompilerServices.AsyncUniTask`1<VoltstroStudios.UnityWebBrowser.Core.WebBrowserClient/<WaitForEngineReadyTask>d__57>:Run () (at Library/PackageCache/com.cysharp.unitask@73d86259ce/Runtime/CompilerServices/StateMachineRunner.cs:189)
Cysharp.Threading.Tasks.AwaiterActions:Continuation (object) (at Library/PackageCache/com.cysharp.unitask@73d86259ce/Runtime/UniTask.cs:21)
Cysharp.Threading.Tasks.UniTaskCompletionSourceCore`1<Cysharp.Threading.Tasks.AsyncUnit>:TrySetException (System.Exception) (at Library/PackageCache/com.cysharp.unitask@73d86259ce/Runtime/UniTaskCompletionSource.cs:167)
Cysharp.Threading.Tasks.CompilerServices.AsyncUniTask`1<Cysharp.Threading.Tasks.UniTaskExtensions/<Timeout>d__12>:SetException (System.Exception) (at Library/PackageCache/com.cysharp.unitask@73d86259ce/Runtime/CompilerServices/StateMachineRunner.cs:210)
Cysharp.Threading.Tasks.UniTaskExtensions/<Timeout>d__12:MoveNext () (at Library/PackageCache/com.cysharp.unitask@73d86259ce/Runtime/UniTaskExtensions.cs:409)
Cysharp.Threading.Tasks.CompilerServices.AsyncUniTask`1<Cysharp.Threading.Tasks.UniTaskExtensions/<Timeout>d__12>:Run () (at Library/PackageCache/com.cysharp.unitask@73d86259ce/Runtime/CompilerServices/StateMachineRunner.cs:189)
Cysharp.Threading.Tasks.AwaiterActions:Continuation (object) (at Library/PackageCache/com.cysharp.unitask@73d86259ce/Runtime/UniTask.cs:21)
Cysharp.Threading.Tasks.UniTaskCompletionSourceCore`1<System.ValueTuple`3<int, bool, bool>>:TrySetResult (System.ValueTuple`3<int, bool, bool>) (at Library/PackageCache/com.cysharp.unitask@73d86259ce/Runtime/UniTaskCompletionSource.cs:139)
Cysharp.Threading.Tasks.UniTask/WhenAnyPromise`2bool, bool:TryInvokeContinuationT2 (Cysharp.Threading.Tasks.UniTask/WhenAnyPromise`2bool, bool,Cysharp.Threading.Tasks.UniTask`1/Awaiter<bool>&) (at Library/PackageCache/com.cysharp.unitask@73d86259ce/Runtime/UniTask.WhenAny.Generated.cs:99)
Cysharp.Threading.Tasks.UniTask/WhenAnyPromise`2/<>c<bool, bool>:.ctorb__2_1 (object) (at Library/PackageCache/com.cysharp.unitask@73d86259ce/Runtime/UniTask.WhenAny.Generated.cs:58)
Cysharp.Threading.Tasks.UniTaskCompletionSourceCore`1(object).TrySetResult (object) (at Library/PackageCache/com.cysharp.unitask@73d86259ce/Runtime/UniTaskCompletionSource.cs:139)
Cysharp.Threading.Tasks.UniTask/DelayPromise:MoveNext () (at Library/PackageCache/com.cysharp.unitask@73d86259ce/Runtime/UniTask.Delay.cs:687)
Cysharp.Threading.Tasks.Internal.PlayerLoopRunner:RunCore () (at Library/PackageCache/com.cysharp.unitask@73d86259ce/Runtime/Internal/PlayerLoopRunner.cs:175)
Cysharp.Threading.Tasks.Internal.PlayerLoopRunner:Update () (at Library/PackageCache/com.cysharp.unitask@73d86259ce/Runtime/Internal/PlayerLoopRunner.cs:145)
Cysharp.Threading.Tasks.Internal.PlayerLoopRunner:Run () (at Library/PackageCache/com.cysharp.unitask@73d86259ce/Runtime/Internal/PlayerLoopRunner.cs:104)
</pre>
</details>


Attempting to resovle this, I've been:

- retracing steps / misc problem solving. 
- tracking down dll's that get rebuilt from step 1
- tracking down dll's that are normally found from UWB's Unity files, when UWB is working (its default installation & windows package) 

Where I have found them, I've manually replaced them with files from step 1, with no success.

Am I preforming the correct steps ( 1 & 2 )?
If (so) Do these errors look at all familar to you?

Thanks for your time

**Reply by Voltstro** [reply]
You haven't setup your dev environment correctly. Your missing the cef bins which are downloaded in the setup process. Make sure you follow the [dev guide](https://projects.voltstro.dev/UnityWebBrowser/articles/dev/dev-guide/) correctly, and the scripts don't error. If you truly believe your environment is correct, run the [download-cef-windows.ps1](https://github.com/Voltstro-Studios/UnityWebBrowser/blob/master/src/DevScripts/download-cef-windows.ps1) script, and when you build, make sure `libcef.dll` is located next to the engine's exe.

You should be using the UnityProject provided in the repo like stated in the docs for development/testing, not adding the dev package to an external project and developing/testing there.

**Reply by bliqqo** [reply]
You were correct, My dev environment was not set up correctly. Upon having rebuilt and reedited my work with the correct environment multiple times, it seems to throw errors after being built (wrong procedure order possibly), or ignore my changes ( as though they weren't included in the build). 

### Attempt order 1:

1. clone git hub project 
2. change scripts with previously discussed changes
3. run setup-all.ps1                                      -                     _( attempt run from Unity => failed run w/ errors below)_ 
4. manual visual studios rebuild of uwb.engine.cef           -                   _( attempt to run from Unity => failed run w/ errors below)_ 

<details>
  <summary>
    <b>[Web Browser]: Error in data loop! System.IO.IOException: Unable to read data from the transport connection: An existing connection was forcibly closed by the remote host.</b>
  </summary><pre>
[Web Browser]: Error in data loop! System.IO.IOException: Unable to read data from the transport connection: An existing connection was forcibly closed by the remote host.

UnityEngine.Logger:LogError (string,object)
VoltstroStudios.UnityWebBrowser.Logging.DefaultUnityWebBrowserLogger:Error (object) (at D:/CleanTestSlate.2/UnityWebBrowser/src/Packages/UnityWebBrowser/Runtime/Logging/DefaultUnityWebBrowserLogger.cs:36)
VoltstroStudios.UnityWebBrowser.Core.WebBrowserClient:PixelDataLoop () (at D:/CleanTestSlate.2/UnityWebBrowser/src/Packages/UnityWebBrowser/Runtime/Core/WebBrowserClient.cs:514)
System.Threading.ThreadHelper:ThreadStart ()
</pre>
</details>


<details>
  <summary>
    <b>[Web Browser]: It appears that the engine process has quit!</b>
  </summary><pre>
[Web Browser]: It appears that the engine process has quit!

UnityEngine.Logger:LogError (string,object)
VoltstroStudios.UnityWebBrowser.Logging.DefaultUnityWebBrowserLogger:Error (object) (at D:/CleanTestSlate.2/UnityWebBrowser/src/Packages/UnityWebBrowser/Runtime/Logging/DefaultUnityWebBrowserLogger.cs:36)
VoltstroStudios.UnityWebBrowser.Core.WebBrowserClient:PixelDataLoop () (at D:/CleanTestSlate.2/UnityWebBrowser/src/Packages/UnityWebBrowser/Runtime/Core/WebBrowserClient.cs:481)
System.Threading.ThreadHelper:ThreadStart ()
</pre>
</details>

### Attempt order 2:

1. clone git hub project 
2. run setup-all.ps1    
3. change scripts with previously discussed changes            -                   _( attempt to run from Unity => success with no sign of my changes)_
5. manual visual studios rebuild of uwb.engine.cef           -                   _( attempt to run from Unity => success with no sign of my changes)_

Is there a certain order to set up the project with these changes?

_I'll keep saying it so long as your still here, really appreciate the continued assistance_

**Reply by Voltstro** [reply]
The order to setup everything is:

1. Clone the repo recursively
2. Run `setup-all.ps1` script with PowerShell
   - Check for errors, make sure there are none
3. Open `src/UnityWebBrowser.UnityProject` with Unity, open the scene, and run it to check that the default build works
4. Do your changes
5. Run the dev script for the project(s) that you are modifying (E.G: `publish-cef-engine-windows.ps1`)
6. Test in the Unity project again

The new errors you have provided indicate that the CEF engine process is crashing. You can build the external projects in your IDE of choice with the configuration set to debug for a no-strip, symbols build. Then run the engine process that is built to `<Project>/bin/Debug` with the command line to see the errors. (*Note, you may be able to run the project through your IDE to attach a debugger and do your debugging like that, but I know that atleast on a Linux dev environment with JetBrain's Rider the engine process will always crash when a debugger is attached*)

You can launch the process with the arguments `-remote-debugging 5555` to use Chromium's remote headless debugger (on a normal browser goto `localhost:5555` to access it). There is also `-log-severity Debug` for debug logging level.

**Reply by bliqqo** [reply]
**Thanks to all your advice, I have found a solution for the time being. (maybe not _the_ solution, but one that works for me)**

>The new errors you have provided indicate that the CEF engine process is crashing. You can build the external projects in your IDE of choice with the configuration set to debug for a no-strip, symbols build. Then run the engine process that is built to <Project>/bin/Debug with the command line to see the errors. (Note, you may be able to run the project through your IDE to attach a debugger and do your debugging like that, but I know that atleast on a Linux dev environment with JetBrain's Rider the engine process will always crash when a debugger is attached)

This lead me to finding my own exceptions/bugs in my handler, that I wasn't handling correctly. Fixing that and the correctly ordering my setup (as you detailed) created a stable environment for me to make the custom changes to UWB's CEF.

Using UwbCefLoadHandler & UwbCefClient I intercept the URL that will throw the error and replace it with a functional URL (without the odd scheme). This might work in zoom's case, I don't know how useful this might be in other situations, although.

My handler is operating but was not solving the ERR_UNKNOWN_URL_SCHEME, despite overriding the allowOSExection. Which is why I tried the URL interception.

>You can tell CEF to allow an unknown URL scheme to attempt execution via the registered OS protocol handler. This is done via the [CefResourceRequestHandler.OnProtocolExecution](https://gitlab.com/Voltstro-Studios/WebBrowser/cefglue/-/blob/master/CefGlue/Classes.Handlers/CefResourceRequestHandler.cs#L265), so you will need to override the handler in UWB's CEF engine (as it's not currently used), and check for allowed URL schemes, then set allowOSExecution to true.

Do you need to manually state the allowed URL scheme's anywhere, or was that purely a suggestion before setting allowOSExecution to be true?

_Could not have gotten this far without your assistance, thank you._

**Reply by Voltstro** [reply]
> Using UwbCefLoadHandler & UwbCefClient I intercept the URL that will throw the error and replace it with a functional URL (without the odd scheme). This might work in zoom's case, I don't know how useful this might be in other situations, although.

If your goal is just to purely open Zoom inside of UWB, rather than allow the desktop app, then yea, using URL redirection should be fine.

> Do you need to manually state the allowed URL scheme's anywhere, or was that purely a suggestion before setting allowOSExecution to be true?

I don't think so? I mainly suggested checking the scheme due to the security warning stated in the CEF docs.

**Comment by LyraHarring** [comment]
The [ERR_UNKNOWN_URL_SCHEME](https://cheapsslweb.com/resources/how-to-resolve-err_unknown_url_scheme-error) happens when your browser doesn’t recognize the start of a URL (like myapp:// instead of http://). This often occurs if the URL uses a custom scheme that the browser can’t handle, or if there’s a typo in the link. To fix it, check the URL for mistakes, disable conflicting browser extensions, or try using a different browser that supports the custom scheme. As, I was also facing such kind of error and found the correct solution by applying different methods.
